pkg/util/portforward.go

```diff
@@ -0,0 +1,221 @@
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/transport/spdy"
+)
+
+type LocalProxyServer interface {
+	Listen() (func(), error)
+}
+
+// PortForwardProtocolV1Name is the subprotocol used for port forwarding.
+const PortForwardProtocolV1Name = "portforward.k8s.io"
+
+func NewRoundRobinLocalProxy(
+	restConfig *rest.Config,
+	podNamespace,
+	podSelector string,
+	targetPort int32) LocalProxyServer {
+	return &roundRobin{
+		restConfig:           restConfig,
+		proxyServerNamespace: podNamespace,
+		podSelector:          podSelector,
+		targetPort:           targetPort,
+		reqId:                0,
+	}
+}
+
+var _ LocalProxyServer = &roundRobin{}
+
+type roundRobin struct {
+	proxyServerNamespace string
+	podSelector          string
+	targetPort           int32
+
+	restConfig *rest.Config
+	reqId      int // TODO: thread-safety
+}
+
+func (r *roundRobin) Listen() (func(), error) {
+	listener, err := net.Listen(
+		"tcp",
+		net.JoinHostPort("localhost", strconv.Itoa(int(r.targetPort))))
+	if err != nil {
+		return nil, fmt.Errorf("unable to create listener: Error %s", err)
+	}
+	go func() {
+		for {
+			conn, err := listener.Accept()
+			if err != nil {
+				if !strings.Contains(strings.ToLower(err.Error()), "use of closed network connection") {
+					runtime.HandleError(fmt.Errorf("error accepting connection on port %d: %v", r.targetPort, err))
+				}
+			}
+			go func() {
+				if err := r.handle(conn); err != nil {
+					runtime.HandleError(fmt.Errorf("error handling connection: %v", err))
+				}
+			}()
+		}
+	}()
+	return func() { listener.Close() }, nil
+
+}
+
+func (r *roundRobin) handle(conn net.Conn) error {
+	transport, upgrader, err := spdy.RoundTripperFor(r.restConfig)
+	if err != nil {
+		return err
+	}
+	nativeClient, err := kubernetes.NewForConfig(r.restConfig)
+	if err != nil {
+		return err
+	}
+	podList, err := nativeClient.CoreV1().
+		Pods(r.proxyServerNamespace).
+		List(context.TODO(), metav1.ListOptions{
+			LabelSelector: r.podSelector,
+		})
+	if err != nil {
+		return err
+	}
+
+	podIdx := rand.Intn(len(podList.Items))
+	pod := podList.Items[podIdx]
+	req := nativeClient.RESTClient().
+		Post().
+		Prefix("api", "v1").
+		Resource("pods").
+		Namespace(r.proxyServerNamespace).
+		Name(pod.Name).
+		SubResource("portforward")
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, "POST", req.URL())
+	streamConn, _, err := dialer.Dial(PortForwardProtocolV1Name)
+	if err != nil {
+		return err
+	}
+	defer streamConn.Close()
+
+	// create error stream
+	headers := http.Header{}
+	headers.Set(v1.StreamType, v1.StreamTypeError)
+	headers.Set(v1.PortHeader, fmt.Sprintf("%d", r.targetPort))
+	headers.Set(v1.PortForwardRequestIDHeader, strconv.Itoa(r.reqId))
+	r.reqId++
```

Comment:
add a todo here also? I think you would need to at least lock here.

Reply:
added a lock